test/CometWrapper.t.sol

```diff
@@ -340,17 +366,147 @@ contract CometWrapperTest is BaseTest, CometMath {
+
+        vm.prank(bob);
+        cometWrapper.approve(alice, type(uint256).max);
+
+        uint256 aliceCometBalance = comet.balanceOf(alice);
+        uint256 bobAssets = cometWrapper.maxWithdraw(bob);
+        uint256 assetsToWithdraw = 987e6;
+        uint256 expectedBobWrapperAssets = bobAssets - assetsToWithdraw;
+        uint256 expectedAliceCometBalance = aliceCometBalance + assetsToWithdraw;
+
+        // Alice withdraws from Bob to herself
+        vm.startPrank(alice);
+        // TODO: investigate rounding by 2???
+        vm.expectEmit(true, true, true, true);
+        emit Withdraw(alice, alice, bob, assetsToWithdraw, cometWrapper.convertToShares(assetsToWithdraw) + 2);
+        cometWrapper.withdraw(assetsToWithdraw, alice, bob);
+        vm.stopPrank();
+
+        assertEq(cometWrapper.totalSupply(), unsigned104(comet.userBasic(wrapperAddress).principal));
+        assertEq(cometWrapper.totalAssets(), comet.balanceOf(wrapperAddress));
+        // TODO: investigate rounding by 2???
+        assertApproxEqAbs(cometWrapper.underlyingBalance(bob), expectedBobWrapperAssets, 2);
+        assertLe(cometWrapper.underlyingBalance(bob), expectedBobWrapperAssets);
+        assertApproxEqAbs(comet.balanceOf(alice), expectedAliceCometBalance, 1);
+        assertLe(comet.balanceOf(alice), expectedAliceCometBalance);
+    }
+
+    // TODO: failing due to allowance bug
+    function test_withdrawUsesAllowances() public {
+        vm.startPrank(alice);
+        comet.allow(wrapperAddress, true);
+        cometWrapper.mint(5_000e6, alice);
+        vm.stopPrank();
+
+        vm.prank(alice);
+        cometWrapper.approve(bob, 2_700e6);
+
+        uint256 sharesToWithdraw = 2_500e6;
+        uint256 assetsToWithdraw = cometWrapper.convertToAssets(sharesToWithdraw);
+
+        vm.startPrank(bob);
+        // Allowances should be updated when withdraw is done
+        assertEq(cometWrapper.allowance(alice, bob), 2_700e6);
+        cometWrapper.withdraw(assetsToWithdraw, bob, alice);
+        assertApproxEqAbs(cometWrapper.balanceOf(alice), 5_000e6 - sharesToWithdraw, 1);
+
+        // Reverts if trying to withdraw again now that allowance is used up
+        vm.expectRevert(CometHelpers.InsufficientAllowance.selector);
+        cometWrapper.withdraw(assetsToWithdraw, bob, alice);
+        vm.stopPrank();
+        assertEq(cometWrapper.allowance(alice, bob), 200e6);
+
+        // Infinite allowance does not decrease allowance
+        vm.prank(bob);
+        cometWrapper.approve(alice, type(uint256).max);
+        assertEq(cometWrapper.allowance(bob, alice), type(uint256).max);
+
+        vm.startPrank(alice);
+        cometWrapper.withdraw(assetsToWithdraw, alice, bob);
+        assertEq(cometWrapper.allowance(bob, alice), type(uint256).max);
+        vm.stopPrank();
     }
 
-    // TODO: test withdraw from and to
-    // TODO: test withdrawing not a max amount
+    // TODO: failing due to allowance bug
```

Comment:
this is the test that would have caught the bug, yeah? 🙌

Reply:
Yeah this is the one